### PR TITLE
[ENH] add gzip scripts for raw folder 

### PR DESCRIPTION
**Pull request by marcobarilari**
Very simple scripts to gzip any `.nii` file a raw folder. Any suggestion on how to make them smarter and how they are organize in the repo struct is welcome.

### CODE REVIEW

**Comment by marcobarilari** [comment]
btw, strangely enough I cannot change it to a draft pr

File: src/workflows/bidsGZipRawFolder.m [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  %
+  % :param optSource: The structure that contains the options set by the user to run the batch
+  %                   workflow for source processing
+  % :type opt: structure
+  % :param keepUnzippedNii: will keep the original ``.nii`` if set to ``true``. Default is false
+  % :type keepUnzippedNii: boolean
+
+  %% input variables default values
+  
+  optSource.dataDir = '/Users/barilari/Desktop/data_temp/raw_temp';
```
Comment (marcobarilari) [inline src/workflows/bidsGZipRawFolder.m:22]:
```suggestion
```

File: src/batches/setBatchGZip.m [hunk 2]
```diff
@@ -0,0 +1,22 @@
+% (C) Copyright 2020 CPP BIDS SPM-pipeline developers
+
+function matlabbatch = setBatchGZip(unzippedNiifiles, keepUnzippedNii)
+  %
+  % Set the batch for GZip the 4D volumes
+  %
+  % USAGE:: 
+  %
+  %   matlabbatch = setGZip(unzippedNiifiles, deleteUnzippedNii)
```
Comment (Remi-Gau) [inline src/batches/setBatchGZip.m:9]:
```suggestion
  %   matlabbatch = setBatchGZip(unzippedNiifiles, keepUnzippedNii)
```

File: src/batches/setBatchGZip.m [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  %
+  % :param unzippedNiifiles: List of volumes to be gzipped
+  % :type unzippedNiifiles: array
+  % :param keepUnzippedNii: Obligatory argument. Boolean to decide to delete the unzipped files
```
Comment (Remi-Gau) [inline src/batches/setBatchGZip.m:13]:
```suggestion
  % :param keepUnzippedNii: Boolean to decide to delete the unzipped files
```

unless the argument is in sqare brackets in the USAGE section then it is obligatory

File: src/batches/setBatchGZip.m [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  % :type keepUnzippedNii: boolean
+  %
+  % :returns: - :matlabbatch: (struct) The matlabbath ready to run the spm job
+
+  matlabbatch{1}.cfg_basicio.file_dir.file_ops.cfg_gzip_files.files = unzippedNiifiles;
```
Comment (Remi-Gau) [inline src/batches/setBatchGZip.m:18]: Something we might have to decide for all functions that create batches.

That they simply "extend" a previous batch that is passed to them as argument:
```
matlabbatchOUT = setBatchFoo(matlabbatchIN, arg1)
```

where `numel(matlabbatchOUT) = numel(matlabbatchIN)  + 1`

The reason why this default way of setting up is that it makes easy to "chain" `setBatch` functions to create a list of batches to then be executed in one single workflow.

The template could look like that.


```
function matlabbatch = setBatchFoo(matlabbatch, argin2, argin3)
  %
  % Short description of what the function does goes here.
  %
  % USAGE::
  %
  %   [argout1, argout2] = setBatchFoo([matlabbatch] [, argin2 == default] [, argin3])
  %
  % :param matlabbatch: 
  % :type matlabbatch: structure
  % :param argin2: optional argument and its default value. And some of the
  %               options can be shown in litteral like ``this`` or ``that``.
  % :type argin2: string
  % :param argin3: (dimension) optional argument
  %
  % :returns:
  % 
  %           - :matlabbatch: (structure)
  %

  printBatchName('nameOfThisBatch');

  if nargin<1
   matlabbatch = [];
  end

  matlabbatch{end + 1}. %
  matlabbatch{end}. % 

end
```

File: src/workflows/bidsGZipRawFolder.m [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+function bidsGZipRawFolder(optSource, keepUnzippedNii)
+
+  %
+  % GZip the nii files in a ``raw`` bids folders from the. It will do it independently of the task.
+  %
+  % USAGE::
+  %
+  %   bidsGZipRawFolder([opt,] ...
+  %                     [deleteUnzippedNii = true,] ...
+  %                     [modalitiesToZip = {'anat', 'func', 'fmap'}])
```
Comment (Remi-Gau) [inline src/workflows/bidsGZipRawFolder.m:12]:
```suggestion
  %   bidsGZipRawFolder(optSource ...
  %                     [, keepUnzippedNii = false])
```

File: src/workflows/bidsGZipRawFolder.m [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  % :param optSource: The structure that contains the options set by the user to run the batch
+  %                   workflow for source processing
+  % :type opt: structure
+  % :param keepUnzippedNii: will keep the original ``.nii`` if set to ``true``. Default is false
+  % :type keepUnzippedNii: boolean
+
+  %% input variables default values
+  
+
+  if nargin < 2 || isempty(deleteUnzippedNii)
```
Comment (Remi-Gau) [inline src/workflows/bidsGZipRawFolder.m:23]:
```suggestion
  if nargin < 2 || isempty(keepUnzippedNii)
```

**Review by Remi-Gau — CHANGES_REQUESTED** [review]
Almost there. :rocket: 

Just realized that an easier function to write a test for is the `checkOptionsSource.m `.

You can clearly inspire yourself from the one written for `checkOptions`. :wink:

File: src/batches/setBatchGZip.m (hunk 4 above)
Comment (Remi-Gau) [reply src/batches/setBatchGZip.m:18]: Another advantage in terms of coding is that it helps to know "Ah it is a `setBatch` function so the first argument is always `matlabbatch`".

Comment (marcobarilari) [reply src/batches/setBatchGZip.m:18]: let me out this in an issue so we remember

**Comment by marcobarilari** [comment]
add a test for `checkOptionSource`

run on my computer and it saved ~500MB on a 3GB dataset so, around the ~15% of space saved. I do not know though if it is a linear equation.

LMK

**Comment by Remi-Gau** [comment]
> add a test for `checkOptionSource`

And it passes and all !!! :-)

Could you run it locally with MOxUnit ? I suspect we will need some doc about that too.

> run on my computer and it saved ~500MB on a 3GB dataset so, around the ~15% of space saved. I do not know though if it is a linear equation.

I doubt it is linear. It very much depends on what you are zipping.

**Review by Remi-Gau — APPROVED** [review]
We are good to go!

Press the big green button when you want.

**Comment by marcobarilari** [comment]
> Could you run it locally with MOxUnit ? I suspect we will need some doc about that too.

You mean the script `runTests`? It does not work because I think I have to set the MOxUnit on my computer and I haven't done it yet. But yes +1 for adding docs about that

**Comment by Remi-Gau** [comment]
Actually to run all the tests you "simply" need to be in the `tests` folder and run `moxunit_runtests`.

But it does require you to install MOxUnit. Easier than it sounds. :-)

**Comment by Remi-Gau** [comment]
@all-contributors please add @marcobarilari for test

**Comment by Remi-Gau** [comment]
let's try again

@all-contributors please add @marcobarilari for test

**Pull request merged**